docs/architecture/adr-012-handlers-redesign.md

```diff
@@ -0,0 +1,598 @@
+
+    /// Stores the given channel_end at a path associated with the port_id and channel_id.
+    fn store_channel(
+        &mut self,
+        port_channel_id: (PortId, ChannelId),
+        channel_end: &ChannelEnd,
+    ) -> Result<(), Error>;
+
+    fn store_next_sequence_send(
+        &mut self,
+        port_channel_id: (PortId, ChannelId),
+        seq: Sequence,
+    ) -> Result<(), Error>;
+
+    fn store_next_sequence_recv(
+        &mut self,
+        port_channel_id: (PortId, ChannelId),
+        seq: Sequence,
+    ) -> Result<(), Error>;
+
+    fn store_next_sequence_ack(
+        &mut self,
+        port_channel_id: (PortId, ChannelId),
+        seq: Sequence,
+    ) -> Result<(), Error>;
+
+    /// Called upon channel identifier creation (Init or Try message processing).
+    /// Increases the counter which keeps track of how many channels have been created.
+    /// Should never fail.
+    fn increase_channel_counter(&mut self);
+
+    /// Ibc events
+    fn emit_ibc_event(event: IbcEvent);
+
+    ////////////////////////////////////////////////////////
+    /* All "read" methods necessary in execution handlers */
+    ////////////////////////////////////////////////////////
+}
+```
+
+### Host based API
+
+ICS24 defines the minimal set of interfaces which must be provided by an IBC enabled blockchain. We therefore define a
+`Host` trait that formalizes and encapsulates these interfaces.
+
+```rust
+pub trait Host {
+    /// An error type that can represent all host errors.
+    type Error;
+
+    /// The Host's key-value store that must provide access to IBC paths. 
+    type KvStore: IbcStore<Self::Error>;
+
+    /// An event logging facility.
+    type EventEmitter: EventEmitter<Event=Event<DefaultIbcTypes>>;
+
+    /// Methods to access the store (ro & rw). 
+    fn store(&self) -> &Self::KvStore;
+    fn store_mut(&mut self) -> &mut Self::KvStore;
+
+    /// Methods to access the event logger (ro & rw).
+    fn event_emitter(&self) -> &Self::EventEmitter;
+    fn event_emitter_mut(&mut self) -> &mut Self::EventEmitter;
+
+    /// Host system data (that is not required to be stored on the KV store)
+    fn current_timestamp(&self) -> IbcTimestamp;
+    fn current_height(&self) -> IbcHeight;
+
+    /* Other methods that cannot be derived from the Store... */
+}
+```
+
+See the Appendix A below for an exhaustive list of all `Host` methods.
+
+This `Host` trait implementation may then be used to derive blanket implementations for all other contexts such as
+`ValidationContext`, `ExecutionContext`, `Ics26Context`, etc. For e.g. ->
+
+```rust
+pub struct HostValidationContext<H> {
+    host: H,
+}
+
+impl<H: Host> HostValidationContext<H> {
+    pub fn new(host: Host) -> Self {
+        Self { host }
+    }
+}
+
+impl<H: Host> ValidationContext for HostValidationContext<H> { /* ... */ }
+```
+
+#### Store
+
+Hosts are expected to implement the `IbcStore` trait that provides access to all IBC paths as defined
```

Comment:
Note: we currently do not differentiate between provable and private stores since `ibc::core::ics24_host::path::Path` doesn't differentiate between provable and private paths.